src/Components/Components/src/ComponentsMetrics.cs

```diff
@@ -0,0 +1,239 @@
+    private readonly Counter<long> _batchException;
+
+    public bool IsNavigationEnabled => _navigationCount.Enabled;
+
+    public bool IsEventDurationEnabled => _eventDuration.Enabled;
+    public bool IsEventExceptionEnabled => _eventException.Enabled;
+
+    public bool IsParametersDurationEnabled => _parametersDuration.Enabled;
+    public bool IsParametersExceptionEnabled => _parametersException.Enabled;
+
+    public bool IsBatchDurationEnabled => _batchDuration.Enabled;
+    public bool IsBatchExceptionEnabled => _batchException.Enabled;
+
+    public ComponentsMetrics(IMeterFactory meterFactory)
+    {
+        Debug.Assert(meterFactory != null);
+
+        _meter = meterFactory.Create(MeterName);
+        _lifeCycleMeter = meterFactory.Create(LifecycleMeterName);
+
+        _navigationCount = _meter.CreateCounter<long>(
+            "aspnetcore.components.navigation",
+            unit: "{route}",
+            description: "Total number of route changes.");
+
+        _eventDuration = _meter.CreateHistogram(
+            "aspnetcore.components.event.duration",
+            unit: "s",
+            description: "Duration of processing browser event.",
+            advice: new InstrumentAdvice<double> { HistogramBucketBoundaries = MetricsConstants.ShortSecondsBucketBoundaries });
+
+        _eventException = _meter.CreateCounter<long>(
+            "aspnetcore.components.event.exceptions",
+            unit: "{exception}",
+            description: "Total number of exceptions during browser event processing.");
+
+        _parametersDuration = _lifeCycleMeter.CreateHistogram(
+            "aspnetcore.components.update_parameters.duration",
+            unit: "s",
+            description: "Duration of processing component parameters.",
+            advice: new InstrumentAdvice<double> { HistogramBucketBoundaries = MetricsConstants.ShortSecondsBucketBoundaries });
+
+        _parametersException = _lifeCycleMeter.CreateCounter<long>(
+            "aspnetcore.components.update_parameters.exceptions",
+            unit: "{exception}",
+            description: "Total number of exceptions during processing component parameters.");
+
+        _batchDuration = _lifeCycleMeter.CreateHistogram(
+            "aspnetcore.components.rendering.batch.duration",
+            unit: "s",
+            description: "Duration of rendering batch.",
+            advice: new InstrumentAdvice<double> { HistogramBucketBoundaries = MetricsConstants.ShortSecondsBucketBoundaries });
+
+        _batchException = _lifeCycleMeter.CreateCounter<long>(
+            "aspnetcore.components.rendering.batch.exceptions",
+            unit: "{exception}",
+            description: "Total number of exceptions during batch rendering.");
+    }
+
+    public void Navigation(string componentType, string route)
+    {
+        var tags = new TagList
+        {
+            { "component.type", componentType ?? "unknown" },
+            { "route", route ?? "unknown" },
```

Comment:
if we don't know the route, is it better to skip the tag all together? I can see it going both ways - it makes it more explicit this way.

Reply:
I'm thinking about running a query on this. Leaving the tag out could mean you need "LEFT JOIN" to not lose the whole activity row. I don't know if this is realistic situation.

Reply:
the 3rd option is tag is present but value is null.  `{ "route", null },` just for completeness